File: mobile/android/AmbulanceEventSocket.kt
package com.example.ambulance

import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import org.json.JSONArray
import org.json.JSONObject

data class AmbulanceMarkerEvent(
    val ambulanceId: String,
    val status: String,
    val lat: Double,
    val lng: Double,
    val heading: Float?,
    val distanceMeters: Int?
)

data class AmbulanceAlertEvent(
    val ambulanceId: String,
    val severity: String,
    val matchType: String,
    val message: String,
    val distanceMeters: Int,
    val timeToImpactSeconds: Int,
    val predictedIntersection: PredictedIntersection?,
    val rerouteSuggestion: RerouteSuggestion?
)

data class PredictedIntersection(
    val lat: Double,
    val lng: Double,
    val ambulanceTimeToIntersectionSeconds: Int
)

data class RerouteSuggestion(
    val reason: String,
    val avoidLat: Double,
    val avoidLng: Double,
    val avoidRadiusMeters: Int,
    val recommendedWithinSeconds: Int
)

class AmbulanceEventSocket(
    private val backendUrl: String,
    private val client: OkHttpClient = OkHttpClient()
) {
    private var webSocket: WebSocket? = null
    private val _markerEvents = MutableSharedFlow<AmbulanceMarkerEvent>(extraBufferCapacity = 16)
    private val _alertEvents = MutableSharedFlow<AmbulanceAlertEvent>(extraBufferCapacity = 16)

    val markerEvents = _markerEvents.asSharedFlow()
    val alertEvents = _alertEvents.asSharedFlow()

    fun connect() {
        val request = Request.Builder()
            .url("$backendUrl/ws/drivers")
            .build()

        webSocket = client.newWebSocket(request, object : WebSocketListener() {
            override fun onMessage(webSocket: WebSocket, text: String) {
                val payload = JSONObject(text)
                when (payload.getString("type")) {
                    "ambulance.marker" -> _markerEvents.tryEmit(
                        AmbulanceMarkerEvent(
                            ambulanceId = payload.getString("ambulanceId"),
                            status = payload.getString("status"),
                            lat = payload.getJSONObject("position").getDouble("lat"),
                            lng = payload.getJSONObject("position").getDouble("lng"),
                            heading = payload.getJSONObject("position")
                                .optDouble("heading")
                                .takeIf { !it.isNaN() }
                                ?.toFloat(),
                            distanceMeters = payload.optInt("distanceMeters")
                                .takeIf { payload.has("distanceMeters") }
                        )
                    )

                    "ambulance.alert" -> _alertEvents.tryEmit(
                        AmbulanceAlertEvent(
                            ambulanceId = payload.getString("ambulanceId"),
                            severity = payload.getString("severity"),
                            matchType = payload.getString("matchType"),
                            message = payload.getString("message"),
                            distanceMeters = payload.getInt("distanceMeters"),
                            timeToImpactSeconds = payload.getInt("timeToImpactSeconds"),
                            predictedIntersection = payload.optJSONObject("predictedIntersection")
                                ?.let { intersection ->
                                    PredictedIntersection(
                                        lat = intersection.getJSONObject("point").getDouble("lat"),
                                        lng = intersection.getJSONObject("point").getDouble("lng"),
                                        ambulanceTimeToIntersectionSeconds = intersection
                                            .getInt("ambulanceTimeToIntersectionSeconds")
                                    )
                                },
                            rerouteSuggestion = payload.optJSONObject("rerouteSuggestion")
                                ?.let { suggestion ->
                                    RerouteSuggestion(
                                        reason = suggestion.getString("reason"),
                                        avoidLat = suggestion.getJSONObject("avoidPoint").getDouble("lat"),
                                        avoidLng = suggestion.getJSONObject("avoidPoint").getDouble("lng"),
                                        avoidRadiusMeters = suggestion.getInt("avoidRadiusMeters"),
                                        recommendedWithinSeconds = suggestion
                                            .getInt("recommendedWithinSeconds")
                                    )
                                }
                        )
                    )
                }
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                // Reconnect policy belongs in production infrastructure code.
            }
        })
    }

    fun registerRoute(
        sessionId: String,
        driverId: String,
        appState: String,
        pushToken: String?,
        currentLat: Double,
        currentLng: Double,
        heading: Float?,
        speedMps: Float?,
        routeId: String,
        polyline: List<com.google.android.gms.maps.model.LatLng>
    ) {
        val routePoints = JSONArray().apply {
            polyline.forEach { point ->
                put(JSONObject().put("lat", point.latitude).put("lng", point.longitude))
            }
        }

        val message = JSONObject()
            .put("type", "driver.register_route")
            .put("sessionId", sessionId)
            .put("driverId", driverId)
            .put("platform", "android")
            .put("appState", appState)
            .put("pushToken", pushToken)
            .put(
                "currentPosition",
                JSONObject()
                    .put("lat", currentLat)
                    .put("lng", currentLng)
                    .put("heading", heading)
                    .put("speedMps", speedMps)
                    .put("timestamp", System.currentTimeMillis())
            )
            .put(
                "route",
                JSONObject()
                    .put("routeId", routeId)
                    .put("polyline", routePoints)
            )

        webSocket?.send(message.toString())
    }

    fun sendProgress(
        sessionId: String,
        appState: String,
        currentLat: Double,
        currentLng: Double,
        heading: Float?,
        speedMps: Float?
    ) {
        val message = JSONObject()
            .put("type", "driver.progress")
            .put("sessionId", sessionId)
            .put("appState", appState)
            .put(
                "currentPosition",
                JSONObject()
                    .put("lat", currentLat)
                    .put("lng", currentLng)
                    .put("heading", heading)
                    .put("speedMps", speedMps)
                    .put("timestamp", System.currentTimeMillis())
            )

        webSocket?.send(message.toString())
    }
}

File: mobile/ios/AmbulanceEventSocket.swift
import Foundation
import CoreLocation

struct AmbulanceMarkerEvent: Decodable {
  struct Position: Decodable {
    let lat: Double
    let lng: Double
    let heading: Double?
  }

  let type: String
  let ambulanceId: String
  let status: String
  let position: Position
  let distanceMeters: Int?
}

struct AmbulanceAlertEvent: Decodable {
  struct PredictedIntersection: Decodable {
    struct Point: Decodable {
      let lat: Double
      let lng: Double
    }

    let point: Point
    let ambulanceTimeToIntersectionSeconds: Int
  }

  struct RerouteSuggestion: Decodable {
    struct AvoidPoint: Decodable {
      let lat: Double
      let lng: Double
    }

    let reason: String
    let avoidPoint: AvoidPoint
    let avoidRadiusMeters: Int
    let recommendedWithinSeconds: Int
  }

  let type: String
  let ambulanceId: String
  let severity: String
  let matchType: String
  let message: String
  let distanceMeters: Int
  let timeToImpactSeconds: Int
  let predictedIntersection: PredictedIntersection?
  let rerouteSuggestion: RerouteSuggestion?
}

final class AmbulanceEventSocket: NSObject {
  private let session = URLSession(configuration: .default)
  private let decoder = JSONDecoder()
  private var task: URLSessionWebSocketTask?

  var onMarker: ((AmbulanceMarkerEvent) -> Void)?
  var onAlert: ((AmbulanceAlertEvent) -> Void)?

  func connect(baseURL: URL) {
    let socketURL = baseURL.appending(path: "ws/drivers")
    task = session.webSocketTask(with: socketURL)
    task?.resume()
    receiveNext()
  }

  func registerRoute(
    sessionId: String,
    driverId: String,
    appState: String,
    pushToken: String?,
    location: CLLocationCoordinate2D,
    heading: CLLocationDirection?,
    speedMps: CLLocationSpeed?,
    routeId: String,
    routePolyline: [CLLocationCoordinate2D]
  ) {
    let routePoints = routePolyline.map { ["lat": $0.latitude, "lng": $0.longitude] }
    var currentPosition: [String: Any] = [
      "lat": location.latitude,
      "lng": location.longitude,
      "timestamp": Int(Date().timeIntervalSince1970 * 1000)
    ]
    if let heading {
      currentPosition["heading"] = heading
    }
    if let speedMps {
      currentPosition["speedMps"] = speedMps
    }

    var payload: [String: Any] = [
      "type": "driver.register_route",
      "sessionId": sessionId,
      "driverId": driverId,
      "platform": "ios",
      "appState": appState,
      "currentPosition": currentPosition,
      "route": [
        "routeId": routeId,
        "polyline": routePoints
      ]
    ]
    if let pushToken {
      payload["pushToken"] = pushToken
    }

    send(json: payload)
  }

  func sendProgress(
    sessionId: String,
    appState: String,
    location: CLLocationCoordinate2D,
    heading: CLLocationDirection?,
    speedMps: CLLocationSpeed?
  ) {
    var currentPosition: [String: Any] = [
      "lat": location.latitude,
      "lng": location.longitude,
      "timestamp": Int(Date().timeIntervalSince1970 * 1000)
    ]
    if let heading {
      currentPosition["heading"] = heading
    }
    if let speedMps {
      currentPosition["speedMps"] = speedMps
    }

    let payload: [String: Any] = [
      "type": "driver.progress",
      "sessionId": sessionId,
      "appState": appState,
      "currentPosition": currentPosition
    ]

    send(json: payload)
  }

  private func receiveNext() {
    task?.receive { [weak self] result in
      guard let self else { return }

      if case let .success(message) = result,
         case let .string(text) = message,
         let data = text.data(using: .utf8) {
        if let marker = try? self.decoder.decode(AmbulanceMarkerEvent.self, from: data),
           marker.type == "ambulance.marker" {
          self.onMarker?(marker)
        } else if let alert = try? self.decoder.decode(AmbulanceAlertEvent.self, from: data),
                  alert.type == "ambulance.alert" {
          self.onAlert?(alert)
        }
      }

      self.receiveNext()
    }
  }

  private func send(json: [String: Any]) {
    guard
      let data = try? JSONSerialization.data(withJSONObject: json),
      let string = String(data: data, encoding: .utf8)
    else {
      return
    }

    task?.send(.string(string)) { _ in }
  }
}
